### PR TITLE
Add fori loop

### PR DESCRIPTION
**Pull request by jeremiecoullon**
I created this branch off of `repo_setup`. It just modifies the code slightly to work for fori_loop. Note that we first need to merge panding fixes in `repo_setup` to this branch before merging `add_fori_loop` to main.

There's a bunch of code duplication compared with the scan version, but as it's a small library it might be nice to leave it as it is (and I don't think there are any more cases to cover beyond `scan` and `fori_loop`.

@zombie-einstein : what do you think? Out of scope or in?

### CODE REVIEW

**Comment by zombie-einstein** [comment]
> I created this branch off of `repo_setup`. It just modifies the code slightly to work for fori_loop. Note that we first need to merge panding fixes in `repo_setup` to this branch before merging `add_fori_loop` to main.
> 
> There's a bunch of code duplication compared with the scan version, but as it's a small library it might be nice to leave it as it is (and I don't think there are any more cases to cover beyond `scan` and `fori_loop`.
> 
> @zombie-einstein : what do you think? Out of scope or in?

Yeah this makes sense, the difference is just how you handle the iteration number right? I think it makes sense to pull out the common code for maintainability and then just have two different usages for the two cases. Happy to do this part if you just wanted to merge the version you have here?

File: jax_tqdm/__init__.py [hunk 1]
```diff
@@ -1 +1 @@
-from jax_tqdm.pbar import progress_bar_scan
+from jax_tqdm.pbar import progress_bar_scan, progress_bar_fori_loop
```
Comment (zombie-einstein) [inline jax_tqdm/__init__.py:1]: Do we want to shorten these function names a bit? Though can leave for another PR.

File: jax_tqdm/pbar.py [hunk 2]
```diff
@@ -93,3 +93,85 @@ def wrapper_progress_bar(carry, x):
...
 
     return _progress_bar_scan
+
+
+def progress_bar_fori_loop(
+    n: int, message: typing.Optional[str] = None
+) -> typing.Callable:
+    """
+    Progress bar for a JAX fori_loop
+    Parameters
```
Comment (zombie-einstein) [inline jax_tqdm/pbar.py:103]: Should have an empty line above

File: jax_tqdm/pbar.py [hunk 3]
```diff
@@ -93,3 +93,85 @@ def wrapper_progress_bar(carry, x):
...
+    ----------
+    n : int
+        Number of iterations.
+    message : str
+        Optional string to prepend to tqdm progress bar.
+    Returns
```
Comment (zombie-einstein) [inline jax_tqdm/pbar.py:109]: Need empty line above

**Review by zombie-einstein — APPROVED** [review]
A few small formatting things but otherwise LGTM if you want to merge after fixing.

**Comment by jeremiecoullon** [comment]
Yeah the only difference is the signature of `fori_loop` and `scan` which only appear in the wrapper at the end of each decorator. 

To refactor the code, my idea would have been to have a helper function (for example `build_tqdm_host(n, message)`) that returns `_update_progress_bar` and `close_tqdm`.

Then we can build the 2 decorators easily:

```python
def progress_bar_fori_loop(n: int, message: typing.Optional[str] = None) -> typing.Callable:
   
   _update_progress_bar, close_tqdm = build_tqdm_host(n, message)

    def _progress_bar_fori_loop(func):
      """Decorator that adds a progress bar to `body_fun` used in `jax.lax.fori_loop`."""

         def wrapper_progress_bar(i, val):
             _update_progress_bar(i)
             result = func(i, val)
             return close_tqdm(result, i)

         return wrapper_progress_bar

     return _progress_bar_fori_loop
```

This follows the functional style of the rest of the library. 

What do you think? Did you have something else in mind?

File: jax_tqdm/pbar.py (hunk 2 above)
Comment (jeremiecoullon) [reply jax_tqdm/pbar.py:103]: you mean 1 empty line above `"Progress bar for a JAX fori_loop"`

Comment (jeremiecoullon) [reply jax_tqdm/pbar.py:103]: Ah or rather above "parameters"

Comment (zombie-einstein) [reply jax_tqdm/pbar.py:104]: Above "parameters"

**Comment by zombie-einstein** [comment]
> Yeah the only difference is the signature of `fori_loop` and `scan` which only appear in the wrapper at the end of each decorator.
> 
> To refactor the code, my idea would have been to have a helper function (for example `build_tqdm_host(n, message)`) that returns `_update_progress_bar` and `close_tqdm`.
> 
> Then we can build the 2 decorators easily:
> 
> ```python
> def progress_bar_fori_loop(n: int, message: typing.Optional[str] = None) -> typing.Callable:
>    
>    _update_progress_bar, close_tqdm = build_tqdm_host(n, message)
> 
>     def _progress_bar_fori_loop(func):
>       """Decorator that adds a progress bar to `body_fun` used in `jax.lax.fori_loop`."""
> 
>          def wrapper_progress_bar(i, val):
>              _update_progress_bar(i)
>              result = func(i, val)
>              return close_tqdm(result, i)
> 
>          return wrapper_progress_bar
> 
>      return _progress_bar_fori_loop
> ```
> 
> This follows the functional style of the rest of the library.
> 
> What do you think? Did you have something else in mind?

No think this is good, other option could be to have a function that produces both wrapper types, but think your solution is more elegant.

**Comment by jeremiecoullon** [comment]
Nice I'll do it now! 

Any thoughts about another actually while we're at it?

**Comment by zombie-einstein** [comment]
> Nice I'll do it now!
> 
> Any thoughts about another name for the progress bars actually while we're at it?

`scan_pbar` and `loop_pbar` or `scan_tqdm` & `loop_tqdm`?

**Comment by jeremiecoullon** [comment]
> > Nice I'll do it now!
> > Any thoughts about another name for the progress bars actually while we're at it?
> 
> `scan_pbar` and `loop_pbar` or `scan_tqdm` & `loop_tqdm`?

I have a slight preference for `scan_tqdm` & `loop_tqdm` as it makes it clear that we're building a wrapper around tqdm (rather than some random progress bar). It also ends up being consistent with the name JAX-tqdm. 

Should we go for this one?

**Comment by zombie-einstein** [comment]
> > > Nice I'll do it now!
> > > Any thoughts about another name for the progress bars actually while we're at it?
> > 
> > 
> > `scan_pbar` and `loop_pbar` or `scan_tqdm` & `loop_tqdm`?
> 
> I have a slight preference for `scan_tqdm` & `loop_tqdm` as it makes it clear that we're building a wrapper around tqdm (rather than some random progress bar). It also ends up being consistent with the name JAX-tqdm.
> 
> Should we go for this one?

Yep, looks good to me!

**Comment by jeremiecoullon** [comment]
Ready to merge I think! 

Let me know if there are some issues left to clean up

File: jax_tqdm/pbar.py [hunk 4]
```diff
@@ -21,6 +21,67 @@ def progress_bar_scan(n: int, message: typing.Optional[str] = None) -> typing.Ca
...
+            _update_progress_bar(i)
+            result = func(i, val)
+            return close_tqdm(result, i)
+
+        return wrapper_progress_bar
+
+    return _loop_tqdm
+
+
+def build_tqdm(n, message):
```
Comment (zombie-einstein) [inline jax_tqdm/pbar.py:80]: Type annotations would be good here

**Review by zombie-einstein — APPROVED** [review]
Looks great, just type hints on the new function to add.

Did you manage to run linting over this? I need to add instructions for usage and add checks to CI. If not I can pull this and make a PR with linting fixes.

**Comment by jeremiecoullon** [comment]
Just added some type annotations!

I've only run Black on all of this. I'm happy with you setting up proper CI and stuff :)

**Comment by zombie-einstein** [comment]
> Just added some type annotations!
> 
> I've only run Black on all of this. I'm happy with you setting up proper CI and stuff :)

Bonza, merge away!
